File: solutions/090.subsets-ii/subsets-ii.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
const subsetsWithDup = nums => {
  let set = new Set()

  const bfs = (temp, pos) => {
    let item = JSON.stringify(temp.sort())
    if (set.has(item)) return
    set.add(item)

    for (let i = pos; i < nums.length; i++) {
      bfs([...temp, nums[i]], i + 1)
    }
  }

  bfs([], 0)

  return [...set].map(JSON.parse)
};

File: solutions/090.subsets-ii/subsets-ii.py
class Solution:
    def subsetsWithDup(self, nums):
        ret = []
        nums.sort()

        def traverse(pos, tmp):
            nonlocal nums, ret

            ret.append(tmp)

            for i in range(pos, len(nums)):
                if i > pos and nums[i-1] == nums[i]:
                    continue

                traverse(i + 1, tmp + [nums[i]])


        traverse(0, [])
        return ret
